docs/gitbook/tutorials/skipper-progressive-delivery.md

```diff
@@ -0,0 +1,382 @@
+the canary is scaled to zero and the rollout is marked as failed.
+
+```text
+kubectl -n flagger-system logs deploy/flagger -f | jq .msg
+
+"New revision detected! Scaling up podinfo.test"
+"GetRoutes primaryWeight: 100, canaryWeight: 0"
+"canary deployment podinfo.test not ready: waiting for rollout to finish: 0 of 1 updated replicas are available"
+"GetRoutes primaryWeight: 100, canaryWeight: 0"
+"Starting canary analysis for podinfo.test"
+"Pre-rollout check acceptance-test passed"
+"primaryWeight: 95, canaryWeight: 5"
+"Advance podinfo.test canary weight 5"
+"GetRoutes primaryWeight: 95, canaryWeight: 5"
+"primaryWeight: 90, canaryWeight: 10"
+"Advance podinfo.test canary weight 10"
+"GetRoutes primaryWeight: 90, canaryWeight: 10"
+"primaryWeight: 85, canaryWeight: 15"
+"Advance podinfo.test canary weight 15"
+"GetRoutes primaryWeight: 85, canaryWeight: 15"
+"primaryWeight: 80, canaryWeight: 20"
+"Advance podinfo.test canary weight 20"
+"GetRoutes primaryWeight: 80, canaryWeight: 20"
+"Halt podinfo.test advancement success rate 53.42% < 99%"
+"GetRoutes primaryWeight: 80, canaryWeight: 20"
+"Halt podinfo.test advancement success rate 53.19% < 99%"
+"GetRoutes primaryWeight: 80, canaryWeight: 20"
+"Halt podinfo.test advancement success rate 48.05% < 99%"
+"GetRoutes primaryWeight: 80, canaryWeight: 20"
+"Halt podinfo.test advancement success rate 47.62% < 99%"
+"GetRoutes primaryWeight: 80, canaryWeight: 20"
+"Halt podinfo.test advancement success rate 51.90% < 99%"
+"GetRoutes primaryWeight: 80, canaryWeight: 20"
+"Halt podinfo.test advancement success rate 48.84% < 99%"
+"GetRoutes primaryWeight: 80, canaryWeight: 20"
+"Halt podinfo.test advancement success rate 49.32% < 99%"
+"GetRoutes primaryWeight: 80, canaryWeight: 20"
+"Halt podinfo.test advancement success rate 58.33% < 99%"
+"GetRoutes primaryWeight: 80, canaryWeight: 20"
+"Halt podinfo.test advancement success rate 42.53% < 99%"
+"GetRoutes primaryWeight: 80, canaryWeight: 20"
+"Halt podinfo.test advancement success rate 51.72% < 99%"
+"GetRoutes primaryWeight: 80, canaryWeight: 20"
+"Rolling back podinfo.test failed checks threshold reached 10"
+"primaryWeight: 100, canaryWeight: 0"
+"Canary failed! Scaling down podinfo.test"
+```
+
+## Custom metrics
+
+The canary analysis can be extended with Prometheus queries.
+
+The demo app is instrumented with Prometheus so you can create a custom check that will use the
+HTTP request duration histogram to validate the canary.
+
+Create a metric template and apply it on the cluster:
+
+```yaml
+apiVersion: flagger.app/v1beta1
+kind: MetricTemplate
+metadata:
+  name: latency
+  namespace: test
+spec:
+  provider:
+    type: prometheus
+    address: http://flagger-prometheus.skipper:9090
+  query: |
+    histogram_quantile(0.99,
+      sum(
+        rate(
+          skipper_serve_route_duration_seconds_bucket{
+            route=~"{{ $route }}",,
```

Comment:
There is no `$route` in Flagger's vars, can you please add a working example?

Reply:
thanks :+1:  - forgot to remove the var

trying a fixed version with the e2e now...

Reply:
@stefanprodan thanks - adjusted now

We removed templating now completely because skipper encods service/ingress names (replaces non word characters with _).

Found no way to do this using go templating. For some projects we are using sprig instead: https://github.com/Masterminds/sprig

Would you be in favor to extend flagger templating with these functions in a new PR?

Reply:
Yes I think that would be a great addition to Flagger's templates 💯